cpp/src/arrow/dataset/scanner.h

```diff
@@ -37,6 +38,7 @@ namespace dataset {
 /// \brief Shared state for a Scan operation
 struct ARROW_DS_EXPORT ScanContext {
   MemoryPool* pool = arrow::default_memory_pool();
+  internal::ThreadPool* thread_pool = arrow::internal::GetCpuThreadPool();
```

Comment:
Maybe it's time to expose a common `ResourceContext` class that has a MemoryPool and a ThreadPool?